lib/src/main/scala/code/chymyst/jc/Molecules.scala

```diff
@@ -288,51 +295,82 @@ class M[T](val name: String) extends (T => Unit) with NonblockingMolecule[T] {
-  *                     mistake in chemistry.
-  * replyTimeout Will be set to "true" if the molecule was emitted with a timeout and the timeout was reached.
-  * replyRepeated Will be set to "true" if the molecule received a reply more than once.
+  * @param message Error message (showing which other molecules did not receive replies, or received multiple replies).
+  */
+private[jc] final case class ErrorNoReply(message: String) extends ReplyStatus
+
+private[jc] case object ReplyTimedOut extends ReplyStatus {
+  override def isTimedOut: Boolean = true
+}
+
+private[jc] case object ReplyWasRepeated extends ReplyStatus {
+  override def isRepeated: Boolean = true
+}
+/** Indicates that the reaction body correctly replied to the blocking molecule.
+  *
+  * @param result The reply value.
+  * @tparam R Type of the reply value.
+  */
+private[jc] final case class HaveReply[R](result: R) extends ReplyStatus {
+  override def haveReply: Boolean = true
+}
 
+/** This trait contains the implementations of most methods for the [[ReplyValue]] and [[EmptyReplyValue]] classes.
+  *
+  *
+  * semaphore
+  * semaphoreForReplyStatus
   *
   * @tparam T Type of the value that the molecule carries.
   * @tparam R Type of the reply value.
   */
 private[jc] trait AbsReplyValue[T, R] {
 
-  @volatile var result: Option[R] = None
-
-  @volatile private var semaphore: Semaphore = {
-    val s = new Semaphore(0, false); s.drainPermits(); s
-  }
+  @volatile var replyStatus: ReplyStatus = WaitingForReply
 
-  @volatile var errorMessage: Option[String] = None
-
-  @volatile var replyTimedOut: Boolean = false
-
-  @volatile var replyWasRepeated: Boolean = false
-
-  private[jc] def releaseSemaphore(): Unit = synchronized {
-    if (Option(semaphore).isDefined) semaphore.release()
-  }
+  /** This is initialized only once when creating an instance of this
+    * class. The semaphore will be acquired when emitting the molecule and released by the "reply"
+    * action. The semaphore will never be used again once a reply is received.
+    */
+  private val semaphore = new Semaphore(0, false)
 
-  private[jc] def deleteSemaphore(): Unit = synchronized {
-    semaphore = null
-  }
+  /** This is used by the reaction that replies to the blocking molecule, in order to obtain
+    * the reply status safely (without race conditions).
+    * Initially the semaphore has 1 permit. The reaction that replies will use 2 permits.
+    */
+  private val semaphoreForReplyStatus = new Semaphore(0, false)
 
   private[jc] def acquireSemaphore(timeoutNanos: Option[Long]): Boolean =
-    if (Option(semaphore).isDefined)
-      timeoutNanos match {
-        case Some(nanos) => semaphore.tryAcquire(nanos, TimeUnit.NANOSECONDS)
-        case None => semaphore.acquire(); true
-      }
-    else false
+    timeoutNanos match {
+      case Some(nanos) => semaphore.tryAcquire(nanos, TimeUnit.NANOSECONDS)
```

Comment:
I don't think this semaphore actually loads the CPU at all (when the program deadlocks, I don't see CPU usage, but maybe I'm mistaken?). So I wouldn't call it a spinlock.